rpcs/ledgerService_test.go

```diff
@@ -0,0 +1,172 @@
+	return nil
+}
+
+func TestLedgerService(t *testing.T) {
+	genesisID := "testGenesisID"
+	cfg := config.GetDefaultLocal()
+	l := fakeLedger{Mock: &mock.Mock{}}
+	fnet := fakeNetwork{router: mux.NewRouter(), Mock: &mock.Mock{}}
+
+	// Test LedgerService not enabled
+	cfg.EnableLedgerService = false
+	ledgerService := MakeLedgerService(cfg, &l, &fnet, genesisID)
+	fnet.AssertNotCalled(t, "RegisterHTTPHandler", LedgerServiceLedgerPath, ledgerService)
+	ledgerService.Start()
+	require.Equal(t, int32(0), ledgerService.running)
+
+	// Test GET 404
+	rr := httptest.NewRecorder()
+	req, err := http.NewRequest("GET", "/", nil)
+	require.NoError(t, err)
+	ledgerService.ServeHTTP(rr, req)
+	require.Equal(t, http.StatusNotFound, rr.Code)
+
+	// Test LedgerService enabled
+	cfg.EnableLedgerService = true
+	fnet.On("RegisterHTTPHandler", LedgerServiceLedgerPath, mock.Anything).Return()
+	ledgerService = MakeLedgerService(cfg, &l, &fnet, genesisID)
+	fnet.AssertCalled(t, "RegisterHTTPHandler", LedgerServiceLedgerPath, ledgerService)
+	ledgerService.Start()
+	require.Equal(t, int32(1), ledgerService.running)
+
+	// Test GET 400 Bad Version String
+	rr = httptest.NewRecorder()
+	req, err = http.NewRequest("GET", "/v2/foobar/ledger/23", nil)
+	require.NoError(t, err)
+	fnet.router.ServeHTTP(rr, req)
+	require.Equal(t, http.StatusBadRequest, rr.Code)
+	require.Contains(t, rr.Body.String(), "unsupported version '2'")
+
+	// Test Get 400 Bad Genesis ID
+	rr = httptest.NewRecorder()
+	req, err = http.NewRequest("GET", "/v1/foobar/ledger/23", nil)
+	require.NoError(t, err)
+	fnet.router.ServeHTTP(rr, req)
+	require.Equal(t, http.StatusBadRequest, rr.Code)
+	require.Contains(t, rr.Body.String(), "mismatching genesisID 'foobar'")
+
+	// Test Get 400 No Path Vars
+	rr = httptest.NewRecorder()
+	req, err = http.NewRequest("GET", "", nil)
+	require.NoError(t, err)
+	ledgerService.ServeHTTP(rr, req)
+	require.Equal(t, http.StatusBadRequest, rr.Code)
+	require.Contains(t, rr.Body.String(), "missing genesisID")
+
+	// Not Testing non path var handling because I'm not convinced it's reachable given `LedgerServiceLedgerPath`
+
+	// Test Get 400 round out of range
+	rr = httptest.NewRecorder()
+	req, err = http.NewRequest("GET", fmt.Sprintf("/v1/%s/ledger/zzzzzzzzzzzzzzzzzzzzzzz", genesisID), nil)
+	require.NoError(t, err)
+	fnet.router.ServeHTTP(rr, req)
+	require.Equal(t, http.StatusBadRequest, rr.Code)
+	require.Contains(t, rr.Body.String(), "specified round number could not be parsed using base 36 : ")
+
+	// Test Get Catchpoint Not Found
+	rr = httptest.NewRecorder()
+	rnd := 1111
+	b36Rnd, err := strconv.ParseUint(fmt.Sprintf("%d", rnd), 36, 64)
+	require.NoError(t, err)
+	req, err = http.NewRequest("GET", fmt.Sprintf("/v1/%s/ledger/%d", genesisID, rnd), nil)
+	require.NoError(t, err)
+	rcs := readCloseSizer{Mock: &mock.Mock{}}
+	gcp := l.On("GetCatchpointStream", basics.Round(b36Rnd)).Return(&rcs, ledgercore.ErrNoEntry{Round: basics.Round(rnd)})
```

Comment:
Why do you use `On` and also implement the mock `GetCatchpointStream`?

Reply:
Oh this is just asking about the library, I don't know how to use it properly. I thought passing in the function name lets you instantiate an interface so I was wondering when you'd implement the mock vs using things like `l.On`. Is this an assertion?

Reply:
The `On` call sets expectations on the mock--i.e. "I expect "GetCatchpointStream" to be called with this round as an argument. The `Return` method records the desired response. Implementing the method I'm mocking just wires it all up so that the arguments are recorded and the desired response is returned.

If the mock here gets anything other than `basics.Round(b36Rnd)` it would fail. It also allows me to return an error sometimes and a nil error elsewhere.